### PR TITLE
Refs #27452 -- Added Default expression for use by serial fields.

### PR DESCRIPTION
**Pull request by InvalidInterrupt**
https://code.djangoproject.com/ticket/27452

### CODE REVIEW

**Comment by codingjoe** [comment]
Just for reference, this depends on #11783

@InvalidInterrupt, I'd love your help in for of a review on that one too. Thanks for your support on the whole topic, it's much appreciated!

**Comment by InvalidInterrupt** [comment]
@codingjoe Could you point out what part of #11783 is necessary for this; I can't quite figure that out. It seems to me all planned uses of this API already function as I'd expect, but if there's something I'm missing I want to add a test case.

**Comment by codingjoe** [comment]
@InvalidInterrupt sorry, my statement was a bit misleading. It's not dependent, where you can't actually use the expression anywhere unless #11783 is implemented. If you consider this to be just an internal refactoring, your patch should be sufficient.
However, I wanted this be be used as a default field value not only for the `AutoField` but other fields too.

As is the patch doesn't seem to provide any new features.

Here's my suggestions, we wait for the other one, rebase and add the documentation again as well as a test that implements the default in another field than the `AutoField`. I'll do my best, to finish the other patch asap.

Sounds good?
Best, Joe

**Comment by InvalidInterrupt** [comment]
@codingjoe If I'm not mistaken, there's currently no way to provide a `default` for use by the database engine (it's sometimes used temporarily for things like adding a field tough); this is what would be needed to make the `Default` expression generally useful to users I think. Your PR seems to be focused on allowing expression defaults to work in DDL operations, and my experience is that I needed to explicitly _avoid_ passing through the `Default` expression to the database in those cases, because `ALTER TABLE foo ADD COLUMN bar ... DEFAULT DEFAULT;` makes no sense.

**Comment by codingjoe** [comment]
I am not sure I can completely follow you. Would you mind expanding a bit more on what you are trying to convey?

**Comment by InvalidInterrupt** [comment]
@codingjoe It seems like what you're hoping for is that eventually a user could define a field like `expiry = DateTimeField(default=Default(), db_default=Now()+timedelta(hours=2))`; please correct me if this is not the case. To my knowledge Django currently does not have anything for users' field definitions like that hypothetical `db_default` parameter, and your PR's docs claim that you have not implemented it there. I think something providing the functionality of that `db_default` is what we would need before Django users would likely have much use for the `Default` expression.
Your PR allows Django to accept expressions as DB default used temporarily when adding or altering a field; the `Default` expression can't be used like that anyway because the `DEFAULT` keyword isn't a valid value for the column default in the database.
(If this still doesn't make sense please ask again or ask more questions; I feel like I'm having trouble explaining it concisely)

**Review by codingjoe — APPROVED** [review]
Hi @InvalidInterrupt,

Thanks for being so patient with me. Yes, you are right, I was getting ahead of myself. I am actually implementing the feature you describe, but this can stand on its own. I don't see a reason to add documentation for the expression at this point. Furthermore, its behavior is tested well in integration tests, a separate unit test for those return values would not provide any additional value IMHO. I'll mark it for check-in. Thanks again!

Best,
Joe

**Comment by codingjoe** [comment]
Wait, actually for me to do that, I would be great if you could open a new ticket. cc me, so I can accept it and move things ahead.

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -803,6 +803,22 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
...
+        return 'DEFAULT'
+
+    def as_sql(self, compiler, connection):
+        return 'DEFAULT', []
+
+    def as_sqlite(self, compiler, connection):
+        return 'NULL', []
+
+    def as_oracle(self, compiler, connection):
+        return 'NULL', []
```
Comment (LilyFoote) [inline django/db/models/expressions.py:819]: I think these will do the wrong thing if the column is nullable and has a default, unless SQLite and Oracle both treat the presence of a `NULL` as an instruction to use the default value. If that is the case, it's worth a comment to document the behaviour.

**Comment by felixxm** [comment]
IMO, it's no longer needed since 7414704e88d73dafbcfbb85f9bc54cb6111439d3 introduced the `DatabaseDefault` expression.

**Pull request closed without merging**